module/os/freebsd/zfs/kmod_core.c

```diff
@@ -185,18 +185,14 @@ zfsdev_close(void *data)
 	zfsdev_state_t *zs = data;
 
 	ASSERT(zs != NULL);
+	ASSERT3S(zs->zs_minor, >, 0);
 
-	mutex_enter(&zfsdev_state_lock);
-
-	ASSERT(zs->zs_minor != 0);
-
-	zs->zs_minor = -1;
 	zfs_onexit_destroy(zs->zs_onexit);
 	zfs_zevent_destroy(zs->zs_zevent);
 	zs->zs_onexit = NULL;
 	zs->zs_zevent = NULL;
-
-	mutex_exit(&zfsdev_state_lock);
+	membar_producer();
+	zs->zs_minor = -1;
```

Comment:
Random observer comment. Changing zs_minor to -1 last makes zfsdev_get_state_impl() return a partial destroyed object. Are you sure it is safe?

Reply:
Ah you're right, thanks.

Reply:
The destructor is called when the last device file reference is dropped.  What would call zfsdev_get_state_impl() for minor in that state?

Reply:
I'm wondering that too, because I'm not seeing how zfsdev_get_state_impl() is safe either way if other threads are still accessing state without locking.

Reply:
I think all zfsdev_get_state_impl() calls which request any zfsdev_state_type are only checking for equality to NULL, the structure is never freed. So existing lock free accesses seem pretty safe to me, but you may need someone more authoritative than random observer ;)
Changing lock-free implementation is a minefield in my experience.

Reply:
It seems to me unless concurrent lookup and destroy is implicitly not possible, there is easily a chance that `zfsdev_get_state_impl()` finds the state we are about to destroy and returns it, then we destroy it, then the destroyed state gets used (for example by `zfs_onexit_add_cb()`). It looks like all the paths to `zfsdev_get_state_impl()` should not be able to run concurrently with the file descriptor being closed, but I agree let's confirm this.